bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookie/LastMarkCommand.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.bookkeeper.tools.cli.commands.bookie;
+
+import com.beust.jcommander.Parameters;
+import com.google.common.collect.Lists;
+import java.util.List;
+import org.apache.bookkeeper.bookie.Journal;
+import org.apache.bookkeeper.bookie.LedgerDirsManager;
+import org.apache.bookkeeper.bookie.LogMark;
+import org.apache.bookkeeper.conf.ServerConfiguration;
+import org.apache.bookkeeper.tools.cli.helpers.BookieCommand;
+import org.apache.bookkeeper.util.DiskChecker;
+
+/**
+ * A bookie command to print the last log marker.
+ */
+@Parameters(commandDescription = "Print last log marker")
+public class LastMarkCommand extends BookieCommand {
```

Comment:
cguttapalem@cguttapale-ltm1:~/Workspace/Community/bookkeeper/bin$ ./bookkeeper-cli bookie

Usage: bookie-shell bookie [options] [command] [command options]
  Commands:
    lastmark      Print last log marker
      Usage: lastmark [options]

------------------------------------

in the last line of the output, there shouldn't be "[options]" in "Usage" line, since there are no options for "lastmark" command

Reply:
@reddycharan this output is managed by jcommander. I don't think it is easy to get round at this moment. is this a real concern?

Reply:
cguttapalem@cguttapale-ltm1:~/Workspace/Community/bookkeeper/bin$ ./bookkeeper-cli 

Usage: bookie-shell [options] [command] [command options]
  Options:
    -c, --conf
       Bookie Configuration File
    -h, --help
       Show this help message
       Default: false
  Commands:
    bookie      Commands on operating a single bookie
      Usage: bookie [options]

    client      Commands that interact with a cluster
      Usage: client [options]

    cluster      Commands that operate a cluster
      Usage: cluster [options]

    metadata      Commands that interact with metadata storage
      Usage: metadata [options]

--------------------------------------------------------------------
1) it is definitely not obvious for the first time users

2) In the following line what does "bookie-shell" refer to? because when I tried "./bookkeeper-cli bookie-shell bookie" it doesn't output just "bookie" command  usage.

Usage: bookie-shell [options] [command] [command options]

3) in the existing shell we have couple of options for commands - ledgeridformat and entryformat. It seems it is missing in the new CLI

4) I'm concerned about flexibility with jcommander framework for commands with complicated options like 'readlog' command - (readlog      [-msg] <entry_log_id | entry_log_file_name> [-ledgerid <ledgerid> [-entryid <entryid>]] [-startpos <startEntryLogBytePos> [-endpos <endEntryLogBytePos>]])

Reply:
5. What is the expected output for the command "./bookkeeper-cli --help bookie" ? isn't it supposed to print help message just for bookie command?

It is printing the full output of bookkeeper-cli

cguttapalem@cguttapale-ltm1:~/Workspace/Community/bookkeeper/bin$ ./bookkeeper-cli --help bookie
 
Usage: bookie-shell [options] [command] [command options]
  Options:
    -c, --conf
       Bookie Configuration File
    -h, --help
       Show this help message
       Default: false
  Commands:
    bookie      Commands on operating a single bookie
      Usage: bookie [options]

    client      Commands that interact with a cluster
      Usage: client [options]

    cluster      Commands that operate a cluster
      Usage: cluster [options]

    metadata      Commands that interact with metadata storage
      Usage: metadata [options]

Reply:
6) in the output for the following command - what does "[options]" referred to in the following line

"Usage: bookie-shell client [options] [command] [command options]"

cguttapalem@cguttapale-ltm1:~/Workspace/Community/bookkeeper/bin$ ./bookkeeper-cli client
JAVA_HOME not set, using java from PATH. (/usr/bin/java)
Usage: bookie-shell client [options] [command] [command options]
  Commands:
    simpletest      Simple test to create a ledger and write entries to it.
      Usage: simpletest [options]
        Options:
          -a, --ack-quorum-size
             Ack quorum size (default 2)
             Default: 2
          -e, --ensemble-size
             Ensemble size (default 3)
             Default: 3
          -n, --num-entries
             Entries to write (default 100)
             Default: 100
          -w, --write-quorum-size
             Write quorum size (default 2)
             Default: 2

Reply:
"bookie-shell client [options] [command] [command options]" this is formatted by jcommander.
 
for a given command, it has options (seen as "[options]") and then sub-commands ("[command]"). so here "[options]" is the options of "client" command. but there is no "options" defined for "client" command, so it is not printed in the output. it only prints the "commands" for "client" command: "simpletest"